CatchException and ShowTyping middleware core extensions

garypretty opened the pull request:
@cleemullins @Stevenic These are the first 3 additional middleware extensions we have previously discussed that we were waiting for the M2 changes for.  The 3 pieces of middleware are based on those @Stevenic produced, so we should have feature parity when they are added to the JS SDK.

I'll add the timeout warning middleware we discussed on the other thread in another PR once that is ready.

garypretty commented:
@cleemullins Once we have these finalized and in I am happy to start to put together a wiki page to cover the core extensions project and document the usage for what is in there.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        private readonly CatchExceptionHandler _handler;
+
+        public CatchExceptionMiddleware(CatchExceptionHandler handler)
```
Is this a good way to do this?

I wonder if (for example) a virtual method would be cleaner.

garypretty replied:
@cleemullins have you got an example of where we use a virtual method in such a way elsewhere?

garypretty replied:
Updated the CatchExceptionMiddleware to take a generic type of Exception and swapped out the delegate for a func parameter.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,34 @@
+
+        public async Task OnProcessRequest(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            await CatchError(context, next);
```
While I love the clarity, I think the CatchError method code could just be right here.

garypretty replied:
Removed the catch error method and move code directly into OnProcessRequest

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,34 @@
+            {
+                await next();
+            }
+            catch (Exception ex)
```
Perhaps only certain exception types? Should the caller be able to specify? 

.Use( new ExceptionHandler<InvalidOperationException>( (ex)=> { mycode; } );

garypretty replied:
Now that the CatchExceptionMiddleware takes a generic type of Exception, these can be stacked, like so....

```cs
.Use(new CatchExceptionMiddleware<Exception>((context, exception) =>
                {
                    context.SendActivity(context.Request.CreateReply(exception.Message));
                    return Task.CompletedTask;
                }))
                .Use(new CatchExceptionMiddleware<NullReferenceException>((context, exception) =>
                {
                    context.SendActivity(context.Request.CreateReply("Sorry - Null Reference Exception"));
                    return Task.CompletedTask;
                }));
```

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class CatchExceptionMiddleware : IMiddleware
```
How composable should these be? 

Could I have multiple of them, all stacked?

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,55 @@
+            _freqency = frequency;
+        }
+
+        public async Task OnProcessRequest(IBotContext context, MiddlewareSet.NextDelegate next)
```
This is awesome, but the corner cases are making my head hurt. 

I *think* everything you have here is right, and the scope of the timer mirrors the scope of the context.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,55 @@
+        private readonly int _freqency;
+
+        public ShowTypingMiddleware(int delay = 500, int frequency = 2000)
+        {
```
argument / range checking.

garypretty replied:
@cleemullins I am not sure what any argument / range checking gets us here.  The middleware accepts int params which are passed directly to the Timer object which in turn takes int params, each with default values.  I guess we could arbitrarily check that the values are 'reasonable', e.g. not specifying a 10 minute initial delay, but all this does is present the chance of run time errors for the user, whereas left as it is I think the code will compile and run correctly even if no typing indicators are ever sent.  

What were you thinking specifically?

garypretty replied:
@cleemullins ah. wait. non-negative check would be good :) Ill update shortly.

garypretty replied:
@cleemullins checks added to ensure delay is >= 0 and frequency > 0.
Hopefully that should be everything.

cleemullins replied:
perfect.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/CatchException_MiddlewareTests.cs:
```diff
@@ -0,0 +1,44 @@
+
+                    if (context.Request.AsMessageActivity().Text == "error")
+                    {
+                        throw new Exception();
```
Consider putting a well known string on this. "test" or similar

garypretty replied:
Added "test" applicable test strings and now checking for their existence as part of the tests.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/CatchException_MiddlewareTests.cs:
```diff
@@ -0,0 +1,44 @@
+            TestAdapter adapter = new TestAdapter()
+                .Use(new CatchExceptionMiddleware((context, exception) =>
+                {
+                    context.SendActivity(context.Request.CreateReply("Sorry, something went wrong"));
```
Vette that the exception passed in is:
1. The correct type.
2. Has the right description ("test"). Otherwise weird stuff might be going on.

garypretty replied:
Now that the middleware takes a specific generic type of exception and we are checking for the description as part of the tests I think we should be good here.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ShowTyping_MiddlewareTests.cs:
```diff
@@ -0,0 +1,105 @@
+            
+            await new TestFlow(adapter, async (context) =>
+                {
+                    Thread.Sleep(10000);
```
This will sometimes fail, If I understand it correctly. 

Send an indicator every 2000 ms on one thread. 
Have another thread sleep for 10s. 

Every now and then, one will complete before the other. 

Also, consider making this quicker, just so the tests run faster. 10s is a long time for a test! :)

garypretty replied:
I have updated the timings here to avoid the situation you described.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ShowTyping_MiddlewareTests.cs:
```diff
@@ -0,0 +1,105 @@
+
+            await new TestFlow(adapter, async (context) =>
+                {
+                    Thread.Sleep(10000);
```
Similar comment.s

garypretty replied:
Reduced timings accordingly to a couple of seconds max

garypretty commented:
@cleemullins Thanks Chris. I will address the comments tomorrow morning here, so should have this ready to merge shortly after :)

garypretty commented:
@cleemullins I believe I have addressed all of your comments now.  Apologies but I am not clear on how to respond to your comments and reference my specific changes :-| Hopefully all self explanatory though.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,67 @@
+                typingActivityTimer = new Timer(SendTypingTimerCallback, context, _delay, _freqency);
+            }
+
+            await next().ConfigureAwait(false);
```
So if an exception bubble up here you will "leak" the timer. Suggest adding a `try/finally` around this to ensure you always dispose of the `typingActivityTimer`.

garypretty commented:
@drub0y good point re. smaller PRs. Noted for the future 👍

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ActivityFilterMiddleware.cs:
```diff
@@ -0,0 +1,44 @@
+        /// </summary>
+        private readonly ActivityFilterHandler _activityFilterHandler;
+
+        public ActivityFilterMiddleware(string activityType, ActivityFilterHandler handler)
```
Perhaps a code comment, saying activityTypes are found in the ACtivityTypes enum.

cleemullins replied:
I was thinking about this more. 

I *THINK* the "handler" argument *SHOULD* be an IMiddleware component. That would allow composistion of these things, which seems very powerful.

garypretty replied:
How about we add an overload, so either you can use a handler for simple scenarios (e.g. just sending an activity) or IMiddleware for more complex composition scenarios?

garypretty replied:
Overload added with accompanying test - really nice idea @cleemullins!
Code comment re. ActivityTypes enum also added.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ActivityFilterMiddleware.cs:
```diff
@@ -0,0 +1,44 @@
+            {
+                // if the incoming Activity type matches the type of activity we are checking for then
+                // invoke our handler
+                await _activityFilterHandler.Invoke(context, next);
```
This needs the .ConfiguireAwait(false) on it. 

I think.

garypretty replied:
Added.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,31 @@
+                // Continue to route the request through the pipeline
+                // any errors further down the pipeline will be caught by
+                // this try / catch
+                await next();
```
Needs the ConfigureAwait(falsE)

garypretty replied:
Added.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,31 @@
+            catch (T ex)
+            {
+                // If an error is thrown and the exception is of type T then invoke the handler
+                await _handler.Invoke(context, ex);
```
Same.

garypretty replied:
Added.

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,72 @@
+                Type = ActivityTypes.Typing,
+                RelatesTo = context.Request.RelatesTo
+            };
+            context.SendActivity(typingActivity);
```
This is an async method that needs to be awaited.

garypretty replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,72 @@
+            SendTypingActivity(context);
+        }
+
+        private void SendTypingActivity(IBotContext context)
```
This method will need to be async, so it can be awaited.

garypretty replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,72 @@
+        private void SendTypingTimerCallback(object state)
+        {
+            var context = (IBotContext) state;
+            SendTypingActivity(context);
```
This is an async method call, at least at the bottom of the stack. This seems like it should await that.

garypretty replied:
Done

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ActivityFilter_MiddlewareTests.cs:
```diff
@@ -0,0 +1,90 @@
+            TestAdapter adapter = new TestAdapter()
+                .Use(new ActivityFilterMiddleware(ActivityTypes.ConversationUpdate, async (context, next) =>
+                {
+                    await context.SendActivity(context.Request.CreateReply("Handling conversation update"));
```
Consider replacing this with Assert.Fail("Wrong Activity Filter run")), so that if this method is called the test has failed.

garypretty replied:
Done

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ActivityFilter_MiddlewareTests.cs:
```diff
@@ -0,0 +1,90 @@
+            TestAdapter adapter = new TestAdapter()
+                .Use(new ActivityFilterMiddleware(ActivityTypes.Message, async (context, next) =>
+                {
+                    await context.SendActivity(context.Request.CreateReply("Handling a message activity"));
```
Consider having these return seperate text "Handler One", "Handler Two" , so we can tell which one has gone wrong.

garypretty replied:
Amended to Handler 1 / 2 as requested.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ActivityFilter_MiddlewareTests.cs:
```diff
@@ -0,0 +1,90 @@
+                }))
+                .Use(new ActivityFilterMiddleware(ActivityTypes.Message, async (context, next) =>
+                {
+                    await context.SendActivity(context.Request.CreateReply("Handling a message activity"));
```
Note: You can simplify this. context.SendActivity takes a string, so you can just say:
await context.SendActivity("Handler 2");

garypretty replied:
Updated all instances across the 3 middleware components + tests.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ActivityFilter_MiddlewareTests.cs:
```diff
@@ -0,0 +1,90 @@
+namespace Microsoft.Bot.Builder.Core.Extensions.Tests
+{
+    [TestClass]
+    public class ActivityFilter_MiddlewareTests
```
Consider adding a test that registers:
1. Null activity type. Make sure an excpetion is thrown.
2. Empty activity type. Make sure an exception is thrown.

garypretty replied:
Tests added.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/CatchException_MiddlewareTests.cs:
```diff
@@ -0,0 +1,89 @@
+            TestAdapter adapter = new TestAdapter()
+                .Use(new CatchExceptionMiddleware<Exception>((context, exception) =>
+                {
+                    context.SendActivity(context.Request.CreateReply("Generic Exception Caught"));
```
These can all be simplified, as SendActivity takes a string and does the right thing.

garypretty replied:
Updated everywhere.

cleemullins commented on tests/Microsoft.Bot.Builder.Core.Extensions.Tests/ShowTyping_MiddlewareTests.cs:
```diff
@@ -0,0 +1,103 @@
+namespace Microsoft.Bot.Builder.Core.Extensions.Tests
+{
+    [TestClass]
+    public class ShowTyping_MiddlewareTests
```
Consider adding a test that passes in invalid arguments and makes sure the exception is thrown.

garypretty replied:
Tests added for both arguments

garypretty commented:
@cleemullins think that's all sorted now Chris.

garypretty commented:
@cleemullins Can you confirm what's left on this? Think your last comment on Skype went missing 😕

tomlm commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ActivityFilterMiddleware.cs:
```diff
@@ -0,0 +1,65 @@
+            _activityFilterHandler = handler;
+        }
+
+        public ActivityFilterMiddleware(string activityType, IMiddleware nextMiddleware)
```
>string activityType [](start = 40, length = 19)

My preference would be to make this more general, and just have this be a lamda which returns BOOL, so you can do your activity type filter like new ActivityFilterMiddleware((context) => context.Request?.Type == ActivityTypes.Message)

tomlm commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ActivityFilterMiddleware.cs:
```diff
@@ -0,0 +1,65 @@
+        /// </summary>
+        private readonly IMiddleware _nextMiddleware;
+
+        public ActivityFilterMiddleware(string activityType, ActivityFilterHandler handler)
```
>ActivityFilterHandler [](start = 61, length = 21)

I don't see the need for this.  Isn't this just the middleware pipeline?

garypretty replied:
@tomlm Maybe the naming is off here.  The point of this piece of middleware is to allow the developer to take action depending on the incoming activity type. e.g. If there is a conversation update then they might want to send a message back to the user (e.g. a welcome message), similar to the method in the v3 controllers which handled various different activity types. @cleemullins

garypretty replied:
@tomlm for additional context, this is a C# version of the middleware @Stevenic

tomlm commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ActivityFilterMiddleware.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class ActivityFilterMiddleware : IMiddleware
```
>ActivityFilterMiddleware [](start = 17, length = 24)

Then I would just rename this to FilterMiddleware

garypretty replied:
@tomlm Based on the comments above, I do think it is the naming that is throwing things off.  Maybe it should be renamed to something like HandleActivityTypeMiddleware?

tomlm commented on libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class CatchExceptionMiddleware<T> : IMiddleware where T : Exception
```
>CatchExceptionMiddleware [](start = 17, length = 24)

Description of use and benefit?

tomlm replied:
This one is pretty useful and cool

---
In reply to: [174949931](https://github.com/Microsoft/botbuilder-dotnet/pull/283#discussion_r174949931) [](ancestors = 174949931)

garypretty replied:
Description added.

tomlm commented on libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class ShowTypingMiddleware : IMiddleware
```
>ShowTypingMiddleware [](start = 17, length = 20)

I like this one too.  Add comment describing usage?

garypretty replied:
Description added.

cleemullins commented:
@garypretty Hey Gary, I think the Exception and Typing Middleware pieces are good to go. I'm not sure yet of the Activity Matching one is ready, and we may hold off on that for a bit. If you want to pull that one out, I can merge the other two in now.

garypretty commented:
@cleemullins Makes sense. I'll sort that later this evening.

garypretty commented:
@cleemullins I have removed the Activity Filter middleware (I will re-submit this in a separate PR once I have considered the naming etc.). Think we should be good to go with this now for the other two.
